### ISSUE TITLE
Two Moons dataset hyperparameters

### ISSUE THREAD

**Issue by VincentPisztora** [vikasverma1077/ICT#8]
Hi there - Thank you for posting this well organized repo! I might have missed it, but could you share the hyperparameter settings (max consistency coefficient and ramp-up schedule, beta distribution parameter, EMA rate, L2 coefficient, batch size) that were used for the two moons dataset? (Figure 1. in the paper)

Thanks!

**Comment by vikasverma1077** [comment]
this should most likely reproduce the figures:

import numpy as np
import torch
import random

def sample_moons(n_samples=100, noise=0):
    n_per_class = n_samples // 2
    pi = 3.141592653589793

    outer_grid = torch.rand(n_per_class, 1) * pi
    outer_circ_x = torch.cos(outer_grid)
    outer_circ_y = torch.sin(outer_grid)

    inner_grid = torch.rand(n_per_class, 1) * pi
    inner_circ_x = 1 - torch.cos(inner_grid)
    inner_circ_y = 1 - torch.sin(inner_grid) - .5

    X = torch.cat((torch.cat((outer_circ_x, inner_circ_x), 0),
                   torch.cat((outer_circ_y, inner_circ_y), 0)), 1)

    y = torch.cat((torch.zeros(n_per_class),
                   torch.ones(n_per_class)), 0)

    X += torch.randn(X.size()) * noise

    return X.detach(), y.view(-1, 1).detach()


def plot_moons(net, x_labeled, y_labeled, x_unlabeled):
    import matplotlib
    matplotlib.use('Agg')
    import matplotlib.pyplot as plt

    colors = ['blue' if yi == 0 else 'red' for yi in y_labeled]
    plt.scatter(x_unlabeled[:, 0], x_unlabeled[:, 1], c="gray", alpha=.75)
    plt.scatter(x_labeled[:, 0], x_labeled[:, 1], c=colors, s=100)

    x_min, x_max = x_unlabeled[:, 0].min() - 0.25, x_unlabeled[:, 0].max() + 0.25
    y_min, y_max = x_unlabeled[:, 1].min() - 0.25, x_unlabeled[:, 1].max() + 0.25

    xx, yy = np.meshgrid(np.arange(x_min, x_max, .1), np.arange(x_min, x_max, .1))

    Z = net(torch.Tensor(np.c_[xx.ravel(), yy.ravel()])).detach().numpy()
    Z = Z.reshape(xx.shape)

    plt.contour(xx, yy, Z, levels=[0.01, 0.1, .5, 0.9, 0.99], linewidths=[3])

    plt.xlim(x_min, x_max)
    plt.ylim(y_min, y_max)

    plt.savefig('mymoon.png')


def mixup(x, y):
    l = random.uniform(0.01,0.2)
    p = torch.randperm(x.size(0))
    x_mix = l * x + (1 - l) * x[p]
    y_mix = l * y + (1 - l) * y[p]
    return x_mix.detach(), y_mix.detach()


def train_net(x_labeled,
              y_labeled,
              x_unlabeled,
              n_hiddens=128,
              n_iterations=5000,
              lamba=1):

    net = torch.nn.Sequential(
            torch.nn.Linear(x_labeled.size(1), n_hiddens),
            torch.nn.ReLU(),
            torch.nn.Linear(n_hiddens, 1),
            torch.nn.Sigmoid())

    opt = torch.optim.Adam(net.parameters(), lr=1e-2)
    bce = torch.nn.BCELoss()

    for iteration in range(n_iterations):
        opt.zero_grad()
        x_labeled_mix, y_labeled_mix = mixup(x_labeled, y_labeled)
        error_labeled = bce(net(x_labeled), y_labeled)
        error_labeled.backward()

        y_unlabeled = net(x_unlabeled)
        x_unlabeled_mix, y_unlabeled_mix = mixup(x_unlabeled, y_unlabeled)
        error_unlabeled = bce(net(x_unlabeled_mix), torch.round(y_unlabeled_mix))
        (lamba * (error_unlabeled)).backward()
        opt.step()

    return net

if __name__ == "__main__":
    lamba = 1 

    torch.manual_seed(1000)

    x_labeled, y_labeled = sample_moons(6, 0.1)
    x_unlabeled, y_unlabeled = sample_moons(1000, noise=0.1)

    net_pen = train_net(x_labeled, y_labeled, x_unlabeled, lamba=lamba)
    plot_moons(net_pen, x_labeled, y_labeled, x_unlabeled)

**Comment by vikasverma1077** [comment]
this should reproduce the two moons fig https://docs.google.com/document/d/1oRb6so-nXVM2ZVlOeSgHTvUgKmhqoR7TWmFvIUzJtUs/edit?usp=sharing

**Comment by VincentPisztora** [comment]
Thanks for the quick response! I was able to reproduce the results. Just a note - some settings were slightly different than in the paper (num./size of layers, unif distn bounded at .2)
